src/util/codec/mysql/json/path_expr.rs

```diff
@@ -0,0 +1,260 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+// From MySQL 5.7, JSON path expression grammar:
+//     pathExpression ::= scope (pathLeg)*
+//     scope ::= [ columnReference ] '$'
+//     columnReference ::= // omit...
+//     pathLeg ::= member | arrayLocation | '**'
+//     member ::= '.' (keyName | '*')
+//     arrayLocation ::= '[' (non-negative-integer | '*') ']'
+//     keyName ::= ECMAScript-identifier | ECMAScript-string-literal
+//
+// And some implementation limits in MySQL 5.7:
+//     1) columnReference in scope must be empty now;
+//     2) double asterisk(**) could not be last leg;
+//
+// Examples:
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.a') -> "b"
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.c') -> [1, "2"]
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.a', '$.c') -> ["b", [1, "2"]]
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.c[0]') -> 1
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.c[2]') -> NULL
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.c[*]') -> [1, "2"]
+//     select json_extract('{"a": "b", "c": [1, "2"]}', '$.*') -> ["b", [1, "2"]]
+
+// FIXME: remove following later
+#![allow(dead_code)]
+
+use std::ops::Index;
+use std::ascii::AsciiExt;
+use regex::Regex;
+use super::super::Result;
+use super::functions::unquote_string;
+
+pub const PATH_EXPR_ASTERISK: &'static str = "*";
+
+// [a-zA-Z_][a-zA-Z0-9_]* matches any identifier;
+// "[^"\\]*(\\.[^"\\]*)*" matches any string literal which can carry escaped quotes.
+const PATH_EXPR_LEG_RE_STR: &'static str =
+    r#"(\.\s*([a-zA-Z_][a-zA-Z0-9_]*|\*|"[^"\\]*(\\.[^"\\]*)*")|(\[\s*([0-9]+|\*)\s*\])|\*\*)"#;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum PathLeg {
+    /// `Key` indicates the path leg  with '.key'.
+    Key(String),
+    /// `Index` indicates the path leg with form '[number]'.
+    Index(i32),
+    /// `DoubleAsterisk` indicates the path leg with form '**'.
+    DoubleAsterisk,
+}
+
+// ArrayIndexAsterisk is for parsing '*' into a number.
+// we need this number represent "all".
+pub const PATH_EXPR_ARRAY_INDEX_ASTERISK: i32 = -1;
+
+pub type PathExpressionFlag = u8;
+
+pub const PATH_EXPRESSION_CONTAINS_ASTERISK: PathExpressionFlag = 0x01;
+pub const PATH_EXPRESSION_CONTAINS_DOUBLE_ASTERISK: PathExpressionFlag = 0x02;
+
+pub fn contains_any_asterisk(flags: PathExpressionFlag) -> bool {
+    (flags & (PATH_EXPRESSION_CONTAINS_ASTERISK | PATH_EXPRESSION_CONTAINS_DOUBLE_ASTERISK)) != 0
+}
+
+#[derive(Clone, Default, Debug, PartialEq)]
+pub struct PathExpression {
+    pub legs: Vec<PathLeg>,
+    pub flags: PathExpressionFlag,
```

Comment:
What if we use `VecDeque` instead of `Vec` and make `flags` part of `PathLeg`, I guess we may get a better performance with the following actions:
 1. compute all flags for each `PathLeg` with just 1 time instead of `legs.len()` times now.
2. `clone` would only happens when meet `DoubleAsterisk ` in `extract_json` instead of each `extract_json` now.

Do I make myself clear?  I guess we may  have a communication offline?